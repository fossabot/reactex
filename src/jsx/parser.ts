// See https://medium.com/the-guild/implementing-a-runtime-version-of-jsx-78e004bf432e
interface NodeSharedField {
  type: string
  length: number // which represents the length of the sub-string in the code that the node occupies. This will help us trim the code string as we go with the parsing process so we can always focus on relevant parts of the string for the current node:
}

interface ElementNode extends NodeSharedField{
  type: 'element'
  name: string
  props: PropsNode
  children: (ElementNode | ValueNode)[]
}

interface PropsNode extends NodeSharedField {
  type: 'props'
  props: { [key: string]: ValueNode }
}

interface ValueNode extends NodeSharedField {
  type: 'value'
  value: string | boolean | number
}

const parseElement = (str: string) : ElementNode | ValueNode => {
  let match
  let length

  const node : ElementNode = {
    type: 'element',
    props: parseProps(''),
    children: [],
    length: 0,
    name: ''
  }

  match = str.match(/<(\w+)/)

  if(!match){
    str = str.split('<')[0]
    return parseValue(str)
  }

  node.name = match[1]
  length = match.index! + match[0].length
  str = str.slice(length)
  node.length += length

  match = str.match(/>/)
  if(!match)
    return node
  
  node.props = parseProps(str.slice(0, match.index))
  length = node.props.length
  str = str.slice(length)
  node.length += length

  match = str.match(/^ *\/ *>/)
  if(match){
    node.length += match.index! + match[0].length
    return node
  }

  match = str.match(/>/)
  if(!match)
    return node
  
  length = match.index! + 1
  str = str.slice(length)
  node.length += length

  let child = parseElement(str)
  while(child.type === 'element' || (child.type === 'value' && child.value)) {
    length = child.length
    str = str.slice(length)
    node.length += length
    node.children.push(child)
    child = parseElement(str)
  }

  match = str.match(new RegExp(`</${node.name}>`))
  if(!match)
    return node
  
  node.length += match.index! + match[0].length

  return node
}

const parseProps = (str: string) => {
  let match

  const node: PropsNode = {
    type: 'props',
    length: 0,
    props: {
      
    }
  }

  const matchNextProp = () => str.match(/ *\w+="(?:.*[^\\]")?/) || str.match(/ *\w+/)

  match = matchNextProp()

  while(match){
    const propStr = match[0]!
    let [key, ...value] = propStr.split('=')
    node.length += propStr.length
    key = key.trim()
    let valueString = value.join('=')
    node.props[key] = {
      type: 'value',
      value: valueString ? valueString.slice(-1, 1) : true,
      length: 0 // TODO ?
    }
    str = str.slice(0, match.index) + str.slice(match.index! + propStr.length)

    match = matchNextProp()
  }

  return node
}

const parseValue = (str: string) => {
  const node : ValueNode = { 
    type: 'value',
    length: str.length,
    value: str.trim()
  }
  return node
}

export default parseElement