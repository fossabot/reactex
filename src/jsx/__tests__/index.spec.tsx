import React from "react"
import parser from '../parser'
import {jsxtex, reactex } from '../'

test('Parser', () => {
  expect(parser(`<reactex_document>The square of five is written as <reactex_math>5^2</reactex_math></reactex_document>`)).toMatchSnapshot()
})

const toJSON = (o: any) => JSON.stringify(o, null, 2)

test('Simple JSX test', () => {
  expect(toJSON(jsxtex `<reactex_document></reactex_document>`)).toEqual(toJSON(<reactex_document></reactex_document>))
})

test('Simple JSX test', () => {
  expect(toJSON(reactex('<reactex_document></reactex_document>'))).toEqual(toJSON(<reactex_document></reactex_document>))
})

test('More complex JSX test', () => {
  expect(toJSON(jsxtex `<reactex_document>The square of five is written as <reactex_math display key="tmp">5^2</reactex_math></reactex_document>`))
  .toEqual(toJSON(<reactex_document>The square of five is written as <reactex_math display key="tmp">5^2</reactex_math></reactex_document>))
})