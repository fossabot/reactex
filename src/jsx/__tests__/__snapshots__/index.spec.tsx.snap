// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser 1`] = `
Object {
  "children": Array [
    Object {
      "length": 33,
      "type": "value",
      "value": "The square of five is written as ",
    },
    Object {
      "children": Array [
        Object {
          "length": 3,
          "type": "value",
          "value": "5^2",
        },
      ],
      "length": 32,
      "name": "reactex_math",
      "props": Object {
        "length": 0,
        "props": Object {},
        "type": "props",
      },
      "tag": "reactex_math",
      "type": "element",
    },
  ],
  "length": 102,
  "name": "reactex_document",
  "props": Object {
    "length": 0,
    "props": Object {},
    "type": "props",
  },
  "tag": "reactex_document",
  "type": "element",
}
`;
